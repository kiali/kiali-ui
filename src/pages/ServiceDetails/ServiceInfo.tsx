import * as React from 'react';
import { style } from 'typestyle';
import {
  Col,
  Icon,
  Nav,
  NavItem,
  Row,
  TabContainer,
  TabContent,
  ToastNotification,
  ToastNotificationList
} from 'patternfly-react';

import ServiceId from '../../types/ServiceId';
import ServiceInfoDescription from './ServiceInfo/ServiceInfoDescription';
import { ServiceDetailsInfo, severityToIconName, validationToSeverity } from '../../types/ServiceInfo';
import ServiceInfoVirtualServices from './ServiceInfo/ServiceInfoVirtualServices';
import ServiceInfoDestinationRules from './ServiceInfo/ServiceInfoDestinationRules';
import ServiceInfoWorkload from './ServiceInfo/ServiceInfoWorkload';
import { ObjectValidation, Validations } from '../../types/IstioObjects';
import { TabPaneWithErrorBoundary } from '../../components/ErrorBoundary/WithErrorBoundary';
import IstioWizardDropdown from '../../components/IstioWizards/IstioWizardDropdown';
import { ThreeScaleInfo, ThreeScaleServiceRule } from '../../types/ThreeScale';
import { DurationDropdownContainer } from '../../components/DurationDropdown/DurationDropdown';
import RefreshButtonContainer from '../../components/Refresh/RefreshButton';

interface ServiceDetails extends ServiceId {
  serviceDetails: ServiceDetailsInfo;
  gateways: string[];
  validations: Validations;
  onRefresh: () => void;
  onSelectTab: (tabName: string, postHandler?: (tabName: string) => void) => void;
  activeTab: (tabName: string, whenEmpty: string) => string;
  threeScaleInfo: ThreeScaleInfo;
  threeScaleServiceRule?: ThreeScaleServiceRule;
}

type ServiceInfoState = {
  error: boolean;
  errorMessage: string;
};

interface ValidationChecks {
  hasVirtualServiceChecks: boolean;
  hasDestinationRuleChecks: boolean;
}

const tabName = 'list';
const tabIconStyle = style({
  fontSize: '0.9em'
});

class ServiceInfo extends React.Component<ServiceDetails, ServiceInfoState> {
  constructor(props: ServiceDetails) {
    super(props);
    this.state = {
      error: false,
      errorMessage: ''
    };
  }

  validationChecks(): ValidationChecks {
    const validationChecks = {
      hasVirtualServiceChecks: false,
      hasDestinationRuleChecks: false
    };
    const validations = this.props.validations || {};
    validationChecks.hasVirtualServiceChecks = this.props.serviceDetails.virtualServices.items.some(
      virtualService =>
        validations.virtualservice &&
        validations.virtualservice[virtualService.metadata.name] &&
        validations.virtualservice[virtualService.metadata.name].checks &&
        validations.virtualservice[virtualService.metadata.name].checks.length > 0
    );

    validationChecks.hasDestinationRuleChecks = this.props.serviceDetails.destinationRules.items.some(
      destinationRule =>
        validations.destinationrule &&
        destinationRule.metadata &&
        validations.destinationrule[destinationRule.metadata.name] &&
        validations.destinationrule[destinationRule.metadata.name].checks &&
        validations.destinationrule[destinationRule.metadata.name].checks.length > 0
    );

    return validationChecks;
  }

  errorBoundaryMessage(resourceName: string) {
    return `One of the ${resourceName} associated to this service has an invalid format`;
  }

  getServiceValidation(): ObjectValidation {
    if (this.props.validations && this.props.validations.service) {
      return this.props.validations.service[this.props.serviceDetails.service.name];
    }
    return {} as ObjectValidation;
  }

  render() {
    const workloads = this.props.serviceDetails.workloads || [];
    const virtualServices = this.props.serviceDetails.virtualServices || [];
    const destinationRules = this.props.serviceDetails.destinationRules || [];
    const validations = this.props.validations || {};
    const validationChecks = this.validationChecks();
    const getSeverityIcon: any = (severity: string = 'error') => (
      <span className={tabIconStyle}>
        {' '}
        <Icon type="pf" name={severityToIconName(severity)} />
      </span>
    );

    const getValidationIcon = (keys: string[], type: string) => {
      let severity = 'warning';
      keys.forEach(key => {
        const validationsForIcon = (this.props.validations || {})![type][key];
        if (validationToSeverity(validationsForIcon) === 'error') {
          severity = 'error';
        }
      });
      return getSeverityIcon(severity);
    };

    return (
      <div>
        {this.state.error ? (
          <ToastNotificationList>
            <ToastNotification type="danger">
              <span>
                <strong>Error </strong>
                {this.state.errorMessage}
              </span>
            </ToastNotification>
          </ToastNotificationList>
        ) : null}
        <div className="container-fluid container-cards-pf">
          <Row className="row-cards-pf">
            <Col xs={12} sm={12} md={12} lg={12}>
              <span style={{ float: 'right' }}>
                <DurationDropdownContainer id="service-info-duration-dropdown" />{' '}
                <RefreshButtonContainer handleRefresh={this.props.onRefresh} />
                &nbsp;
                <IstioWizardDropdown
                  namespace={this.props.namespace}
                  serviceName={this.props.serviceDetails.service.name}
                  show={false}
                  workloads={workloads}
                  virtualServices={virtualServices}
                  destinationRules={destinationRules}
                  gateways={this.props.gateways}
                  tlsStatus={this.props.serviceDetails.namespaceMTLS}
                  onChange={this.props.onRefresh}
                  threeScaleInfo={this.props.threeScaleInfo}
                  threeScaleServiceRule={this.props.threeScaleServiceRule}
                />
              </span>
            </Col>
          </Row>
          <Row className="row-cards-pf">
            <Col xs={12} sm={12} md={12} lg={12}>
              <ServiceInfoDescription
                name={this.props.serviceDetails.service.name}
                namespace={this.props.namespace}
                createdAt={this.props.serviceDetails.service.createdAt}
                resourceVersion={this.props.serviceDetails.service.resourceVersion}
                istioEnabled={this.props.serviceDetails.istioSidecar}
                labels={this.props.serviceDetails.service.labels}
                selectors={this.props.serviceDetails.service.selectors}
                ports={this.props.serviceDetails.service.ports}
                type={this.props.serviceDetails.service.type}
                ip={this.props.serviceDetails.service.ip}
                endpoints={this.props.serviceDetails.endpoints}
                health={this.props.serviceDetails.health}
                externalName={this.props.serviceDetails.service.externalName}
                threeScaleServiceRule={this.props.threeScaleServiceRule}
                validations={this.getServiceValidation()}
              />
            </Col>
          </Row>
          <Row className="row-cards-pf">
            <Col xs={12} sm={12} md={12} lg={12}>
              <TabContainer
                id="service-tabs"
                activeKey={this.props.activeTab(tabName, 'workloads')}
                onSelect={this.props.onSelectTab(tabName)}
              >
                <div>
                  <Nav bsClass="nav nav-tabs nav-tabs-pf">
                    <NavItem eventKey={'workloads'}>{'Workloads (' + Object.keys(workloads).length + ')'}</NavItem>
                    <NavItem eventKey={'virtualservices'}>
                      {'Virtual Services (' + virtualServices.items.length + ')'}
                      {validationChecks.hasVirtualServiceChecks
                        ? getValidationIcon(
                            (this.props.serviceDetails.virtualServices.items || []).map(a => a.metadata.name),
                            'virtualservice'
                          )
                        : undefined}
                    </NavItem>
                    <NavItem eventKey={'destinationrules'}>
                      {'Destination Rules (' + destinationRules.items.length + ')'}
                      {validationChecks.hasDestinationRuleChecks
                        ? getValidationIcon(
                            (this.props.serviceDetails.destinationRules.items || []).map(a => a.metadata.name),
                            'destinationrule'
                          )
                        : undefined}
                    </NavItem>
                  </Nav>
                  <TabContent>
                    <TabPaneWithErrorBoundary eventKey={'workloads'} message={this.errorBoundaryMessage('Workloads')}>
                      {(Object.keys(workloads).length > 0 || this.props.serviceDetails.istioSidecar) && (
                        <ServiceInfoWorkload workloads={workloads} namespace={this.props.namespace} />
                      )}
                    </TabPaneWithErrorBoundary>
                    <TabPaneWithErrorBoundary
                      eventKey={'virtualservices'}
                      message={this.errorBoundaryMessage('Virtual Services')}
                    >
                      {(virtualServices.items.length > 0 || this.props.serviceDetails.istioSidecar) && (
                        <ServiceInfoVirtualServices
                          virtualServices={virtualServices.items}
                          validations={validations!.virtualservice}
                        />
                      )}
                    </TabPaneWithErrorBoundary>
                    <TabPaneWithErrorBoundary
                      eventKey={'destinationrules'}
                      message={this.errorBoundaryMessage('Destination Rules')}
                    >
                      {(destinationRules.items.length > 0 || this.props.serviceDetails.istioSidecar) && (
                        <ServiceInfoDestinationRules
                          destinationRules={destinationRules.items}
                          validations={validations!.destinationrule}
                        />
                      )}
                    </TabPaneWithErrorBoundary>
                  </TabContent>
                </div>
              </TabContainer>
            </Col>
          </Row>
        </div>
      </div>
    );
  }
}

export default ServiceInfo;
