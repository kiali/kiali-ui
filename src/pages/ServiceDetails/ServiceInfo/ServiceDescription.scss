.small-text-donut .donut-title-big-pf {
  font-size: 18px;
}

.small-text-donut .c3-chart-arc text {
  display: none;
}
