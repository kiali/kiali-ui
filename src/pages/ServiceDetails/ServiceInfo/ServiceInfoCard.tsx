import * as React from 'react';
import { Icon } from 'patternfly-react';

interface CardProps {
  iconType?: string;
  iconName?: string;
  title?: string;
  items?: any;
}

class ServiceInfoCard extends React.Component<CardProps> {
  constructor(props: CardProps) {
    super(props);
  }

  render() {
    return (
      <div className="card-pf">
        <div className="card-pf-heading">
          <h2 className="card-pf-title">
            <Icon type={this.props.iconType} name={this.props.iconName} style={{ marginRight: '10px' }} />
            {this.props.title}
          </h2>
        </div>
        <div className="card-pf-body">{this.props.items}</div>
      </div>
    );
  }
}

export default ServiceInfoCard;
