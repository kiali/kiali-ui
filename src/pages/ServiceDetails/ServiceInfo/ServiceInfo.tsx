import * as React from 'react';
import ServiceId from '../../../types/ServiceId';
import ServiceInfoBadge from './ServiceInfoBadge';
import ServiceInfoCard from './ServiceInfoCard';
import { Endpoints, Pod, Port, Source, Rule } from '../../../types/ServiceInfo';
import * as API from '../../../services/Api';
import { ToastNotification, ToastNotificationList, Col, Row } from 'patternfly-react';

type ServiceInfoState = {
  labels?: Map<string, string>;
  type: string;
  name: string;
  ip: string;
  ports?: Port[];
  endpoints?: Endpoints[];
  pods?: Pod[];
  rules?: Rule[];
  dependencies?: Source[];
  error: boolean;
  errorMessage: string;
};

class ServiceInfo extends React.Component<ServiceId, ServiceInfoState> {
  constructor(props: ServiceId) {
    super(props);
    this.state = {
      labels: new Map(),
      name: '',
      type: '',
      ip: '',
      ports: [],
      pods: [],
      rules: [],
      dependencies: [],
      error: false,
      errorMessage: ''
    };
  }

  componentWillMount() {
    console.log('Fetching info of a service...');
    API.GetServiceDetail(this.props.namespace, this.props.service)
      .then(response => {
        console.log(response['data']);
        let data = response['data'];
        this.setState({
          labels: data.labels,
          name: data.name,
          type: data.type,
          ports: data.ports,
          endpoints: data.endpoints,
          pods: data.pods,
          dependencies: data.dependencies,
          rules: data.route_rules,
          ip: data.ip
        });
      })
      .catch(error => {
        this.setState({
          error: true,
          errorMessage: 'Could not connect to server'
        });
        console.log(error);
      });
  }

  render() {
    return (
      <div>
        {this.state.error ? (
          <ToastNotificationList>
            <ToastNotification type="danger">
              <span>
                <strong>Error </strong>
                {this.state.errorMessage}
              </span>
            </ToastNotification>
          </ToastNotificationList>
        ) : null}
        <div className="container-fluid container-cards-pf">
          <Row className="row-cards-pf">
            <Col xs={12} sm={12} md={12} lg={12}>
              <ServiceInfoCard
                iconType="pf"
                iconName="service"
                title={this.state.name}
                items={
                  <Row>
                    <Col xs={12} sm={6} md={6} lg={6}>
                      <div className="progress-description">
                        <strong>Labels</strong>
                      </div>
                      {Object.keys(this.state.labels || new Map()).map((key, i) => (
                        <ServiceInfoBadge
                          key={'label_' + i}
                          scale={0.8}
                          style="plastic"
                          color="#0088ce"
                          leftText={key}
                          rightText={this.state.labels ? this.state.labels[key] : ''}
                        />
                      ))}
                      <div className="progress-description">
                        <strong>Type</strong> {this.state ? this.state.type : ''} <strong> Ip</strong>{' '}
                        {this.state ? this.state.ip : ''}
                      </div>
                    </Col>
                    <Col xs={12} sm={6} md={6} lg={6}>
                      <div className="progress-description">
                        <strong>Ports</strong>
                      </div>
                      {(this.state.ports || []).map((port, i) => (
                        <span style={{ marginLeft: '10px' }} key={'port_' + i}>
                          {port.protocol} {port.name} ({port.port})
                        </span>
                      ))}
                    </Col>
                    <Col xs={12} sm={6} md={12} lg={12}>
                      <hr />
                      <div className="progress-description">
                        <strong>Endpoints</strong>
                      </div>
                      {(this.state.endpoints || []).map((endpoint, i) => (
                        <Row key={'endpoint_' + i}>
                          <Col xs={12} sm={6} md={6} lg={6}>
                            <ul>
                              <li style={{ listStyleType: 'none' }}>Addresses</li>
                              <ul>
                                {(endpoint.addresses || []).map((address, u) => (
                                  <li key={'endpoint_' + i + '_address_' + u}>
                                    {address.name} ({address.ip})
                                  </li>
                                ))}
                              </ul>
                            </ul>
                          </Col>
                          <Col xs={12} sm={6} md={6} lg={6}>
                            <ul>
                              <li style={{ listStyleType: 'none' }}>Ports</li>
                              <ul>
                                {(endpoint.ports || []).map((port, u) => (
                                  <li key={'endpoint_' + i + '_port_' + u}>
                                    {port.protocol} {port.name} ({port.port})
                                  </li>
                                ))}
                              </ul>
                            </ul>
                          </Col>
                        </Row>
                      ))}
                    </Col>
                  </Row>
                }
              />
            </Col>
            <Col xs={12} sm={6} md={4} lg={4}>
              <ServiceInfoCard
                iconType="fa"
                iconName="cube"
                title="Pods"
                items={(this.state.pods || []).map((pod, u) => (
                  <div key={'pods_' + u}>
                    <div className="progress-description">{pod['Name']}</div>
                    {Object.keys(pod.labels || new Map()).map((key, i) => (
                      <ServiceInfoBadge
                        key={'pod_labels_badge_' + i}
                        scale={0.8}
                        style="plastic"
                        color="green"
                        leftText={key}
                        rightText={pod.labels ? pod.labels[key] : ''}
                      />
                    ))}
                  </div>
                ))}
              />
            </Col>
            <Col xs={12} sm={6} md={4} lg={4}>
              <ServiceInfoCard
                iconType="pf"
                iconName="route"
                title="Dependencies"
                items={Object.keys(this.state.dependencies || new Map()).map((key, u) => (
                  <div key={'dependencies_' + u}>
                    <div className="progress-description">{key}</div>
                    <ul>
                      {(this.state.dependencies ? this.state.dependencies[key] : []).map((dependency, i) => (
                        <li key={'dependencies_' + u + '_dependency_' + i}>{dependency}</li>
                      ))}
                    </ul>
                  </div>
                ))}
              />
            </Col>
            <Col xs={12} sm={6} md={4} lg={4}>
              <ServiceInfoCard
                iconType="pf"
                iconName="settings"
                title="Services Source"
                items={(this.state.rules || []).map((rule, i) => (
                  <ul style={{ listStyleType: 'none' }} key={'rule' + i}>
                    <li>
                      <strong>Destination</strong> : {rule.destination ? rule.destination['name'] : ''}
                    </li>
                    <li>
                      <strong>Precendence</strong> :{rule.precedence}
                    </li>
                    <li>
                      <strong>Route</strong>:
                      <ul>
                        {(rule.route || []).map((label, u) =>
                          Object.keys(label.labels || new Map()).map((key, n) => (
                            <li key={'rule_' + i + '_label_' + u + '_n_' + n}>
                              {key} : {label.labels[key]}
                            </li>
                          ))
                        )}
                      </ul>
                    </li>
                    {<hr />}
                  </ul>
                ))}
              />
            </Col>
          </Row>
        </div>
      </div>
    );
  }
}

export default ServiceInfo;
