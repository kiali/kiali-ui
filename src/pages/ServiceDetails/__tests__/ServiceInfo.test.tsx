import * as React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import * as ActualAPI from '../../../services/Api';
import ServiceInfo from '../ServiceInfo';

jest.mock('../../../services/Api');

const API = require('../../../services/Api') as typeof ActualAPI;

describe('#ServiceInfo render correctly with data', () => {
  it('should render serviceInfo with data', () => {
    return API.getServiceDetail('istio-system', 'reviews', true).then(data => {
      const wrapper = shallow(
        <ServiceInfo
          namespace="istio-system"
          service="reviews"
          serviceDetails={data}
          gateways={[]}
          validations={data.validations}
          onRefresh={jest.fn()}
          onSelectTab={jest.fn()}
          activeTab={jest.fn()}
          threeScaleInfo={{
            enabled: false,
            permissions: {
              create: false,
              update: false,
              delete: false
            }
          }}
        />
      );
      expect(shallowToJson(wrapper)).toBeDefined();
      expect(shallowToJson(wrapper)).toMatchSnapshot();
      expect(wrapper.find('ServiceInfoDescription').length === 1).toBeTruthy();
      expect(wrapper.find('InfoRoutes').length === 1).toBeFalsy();
      expect(wrapper.find('ServiceInfoVirtualServices').length === 1).toBeTruthy();
      expect(wrapper.find('ServiceInfoDestinationRules').length === 1).toBeTruthy();
    });
  });
});
