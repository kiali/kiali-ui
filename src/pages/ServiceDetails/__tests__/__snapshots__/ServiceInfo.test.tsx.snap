// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#ServiceInfo render correctly with data should render serviceInfo with data 1`] = `
<Grid
  gutter="md"
  style={
    Object {
      "margin": "30px",
    }
  }
>
  <GridItem
    span={12}
  >
    <ServiceInfoDescription
      createdAt="2018-06-29T16:43:18+02:00"
      endpoints={
        Array [
          Object {
            "addresses": Array [
              Object {
                "ip": "172.17.0.20",
                "kind": "Pod",
                "name": "reviews-v3-5f5bcb6765-hj46f",
              },
              Object {
                "ip": "172.17.0.21",
                "kind": "Pod",
                "name": "reviews-v2-d896b68c-jnxgm",
              },
              Object {
                "ip": "172.17.0.22",
                "kind": "Pod",
                "name": "reviews-v1-5d6696bcf7-2sls7",
              },
            ],
            "ports": Array [
              Object {
                "name": "http",
                "port": 9080,
                "protocol": "TCP",
              },
            ],
          },
        ]
      }
      externalName="my.database.example.com"
      ip="172.30.196.248"
      istioEnabled={true}
      labels={
        Object {
          "app": "reviews",
        }
      }
      name="reviews"
      namespace="istio-system"
      ports={
        Array [
          Object {
            "name": "http",
            "port": 9080,
            "protocol": "TCP",
          },
        ]
      }
      resourceVersion="2652"
      type="ClusterIP"
    />
  </GridItem>
  <GridItem
    span={12}
  >
    <ParameterizedTabs
      activeTab="workloads"
      defaultTab="workloads"
      id="service-tabs"
      onSelect={[Function]}
      tabMap={
        Object {
          "destinationrules": 2,
          "virtualservices": 1,
          "workloads": 0,
        }
      }
      tabName="list"
    >
      <ForwardRef
        eventKey={0}
        title="Workloads (0)"
      >
        <ErrorBoundaryWithMessage
          message="One of the Workloads associated to this service has an invalid format"
        >
          <ServiceInfoWorkload
            namespace="istio-system"
            service={
              Object {
                "apiDocumentation": Object {
                  "hasSpec": true,
                  "type": "rest",
                },
                "destinationRules": Object {
                  "items": Array [
                    Object {
                      "metadata": Object {
                        "creationTimestamp": "2018-07-02T13:44:01+02:00",
                        "name": "reviews",
                        "resourceVersion": "393061",
                      },
                      "spec": Object {
                        "host": "reviews",
                        "subsets": Array [
                          Object {
                            "labels": Object {
                              "version": "v1",
                            },
                            "name": "v1",
                          },
                          Object {
                            "labels": Object {
                              "version": "v2",
                            },
                            "name": "v2",
                          },
                          Object {
                            "labels": Object {
                              "version": "v3",
                            },
                            "name": "v3",
                          },
                        ],
                        "trafficPolicy": undefined,
                      },
                    },
                  ],
                  "permissions": Object {
                    "create": false,
                    "delete": false,
                    "update": false,
                  },
                },
                "endpoints": Array [
                  Object {
                    "addresses": Array [
                      Object {
                        "ip": "172.17.0.20",
                        "kind": "Pod",
                        "name": "reviews-v3-5f5bcb6765-hj46f",
                      },
                      Object {
                        "ip": "172.17.0.21",
                        "kind": "Pod",
                        "name": "reviews-v2-d896b68c-jnxgm",
                      },
                      Object {
                        "ip": "172.17.0.22",
                        "kind": "Pod",
                        "name": "reviews-v1-5d6696bcf7-2sls7",
                      },
                    ],
                    "ports": Array [
                      Object {
                        "name": "http",
                        "port": 9080,
                        "protocol": "TCP",
                      },
                    ],
                  },
                ],
                "health": undefined,
                "istioSidecar": true,
                "service": Object {
                  "createdAt": "2018-06-29T16:43:18+02:00",
                  "externalName": "my.database.example.com",
                  "ip": "172.30.196.248",
                  "labels": Object {
                    "app": "reviews",
                  },
                  "name": "reviews",
                  "ports": Array [
                    Object {
                      "name": "http",
                      "port": 9080,
                      "protocol": "TCP",
                    },
                  ],
                  "resourceVersion": "2652",
                  "type": "ClusterIP",
                },
                "validations": Object {
                  "destinationrule": Object {
                    "reviews": Object {
                      "checks": Array [
                        Object {
                          "message": "This subset is not found from the host",
                          "path": "spec/subsets[0]/version",
                          "severity": "error",
                        },
                        Object {
                          "message": "This subset is not found from the host",
                          "path": "spec/subsets[1]/version",
                          "severity": "error",
                        },
                      ],
                      "name": "details",
                      "objectType": "destinationrule",
                      "valid": false,
                    },
                  },
                },
                "virtualServices": Object {
                  "items": Array [
                    Object {
                      "metadata": Object {
                        "creationTimestamp": "2018-07-02T13:44:01+02:00",
                        "name": "reviews",
                        "resourceVersion": "393057",
                      },
                      "spec": Object {
                        "gateways": undefined,
                        "hosts": Array [
                          "reviews",
                        ],
                        "http": Array [
                          Object {
                            "route": Array [
                              Object {
                                "destination": Object {
                                  "host": "reviews",
                                  "subset": "v1",
                                },
                              },
                            ],
                          },
                        ],
                        "tcp": undefined,
                      },
                    },
                  ],
                  "permissions": Object {
                    "create": false,
                    "delete": false,
                    "update": false,
                  },
                },
                "workloads": Array [],
              }
            }
            workloads={Array []}
          />
        </ErrorBoundaryWithMessage>
      </ForwardRef>
      <ForwardRef
        eventKey={1}
        title={
          <React.Fragment>
            Virtual Services (
            1
            )
          </React.Fragment>
        }
      >
        <ErrorBoundaryWithMessage
          message="One of the Virtual Services associated to this service has an invalid format"
        >
          <ServiceInfoVirtualServices
            service={
              Object {
                "apiDocumentation": Object {
                  "hasSpec": true,
                  "type": "rest",
                },
                "destinationRules": Object {
                  "items": Array [
                    Object {
                      "metadata": Object {
                        "creationTimestamp": "2018-07-02T13:44:01+02:00",
                        "name": "reviews",
                        "resourceVersion": "393061",
                      },
                      "spec": Object {
                        "host": "reviews",
                        "subsets": Array [
                          Object {
                            "labels": Object {
                              "version": "v1",
                            },
                            "name": "v1",
                          },
                          Object {
                            "labels": Object {
                              "version": "v2",
                            },
                            "name": "v2",
                          },
                          Object {
                            "labels": Object {
                              "version": "v3",
                            },
                            "name": "v3",
                          },
                        ],
                        "trafficPolicy": undefined,
                      },
                    },
                  ],
                  "permissions": Object {
                    "create": false,
                    "delete": false,
                    "update": false,
                  },
                },
                "endpoints": Array [
                  Object {
                    "addresses": Array [
                      Object {
                        "ip": "172.17.0.20",
                        "kind": "Pod",
                        "name": "reviews-v3-5f5bcb6765-hj46f",
                      },
                      Object {
                        "ip": "172.17.0.21",
                        "kind": "Pod",
                        "name": "reviews-v2-d896b68c-jnxgm",
                      },
                      Object {
                        "ip": "172.17.0.22",
                        "kind": "Pod",
                        "name": "reviews-v1-5d6696bcf7-2sls7",
                      },
                    ],
                    "ports": Array [
                      Object {
                        "name": "http",
                        "port": 9080,
                        "protocol": "TCP",
                      },
                    ],
                  },
                ],
                "health": undefined,
                "istioSidecar": true,
                "service": Object {
                  "createdAt": "2018-06-29T16:43:18+02:00",
                  "externalName": "my.database.example.com",
                  "ip": "172.30.196.248",
                  "labels": Object {
                    "app": "reviews",
                  },
                  "name": "reviews",
                  "ports": Array [
                    Object {
                      "name": "http",
                      "port": 9080,
                      "protocol": "TCP",
                    },
                  ],
                  "resourceVersion": "2652",
                  "type": "ClusterIP",
                },
                "validations": Object {
                  "destinationrule": Object {
                    "reviews": Object {
                      "checks": Array [
                        Object {
                          "message": "This subset is not found from the host",
                          "path": "spec/subsets[0]/version",
                          "severity": "error",
                        },
                        Object {
                          "message": "This subset is not found from the host",
                          "path": "spec/subsets[1]/version",
                          "severity": "error",
                        },
                      ],
                      "name": "details",
                      "objectType": "destinationrule",
                      "valid": false,
                    },
                  },
                },
                "virtualServices": Object {
                  "items": Array [
                    Object {
                      "metadata": Object {
                        "creationTimestamp": "2018-07-02T13:44:01+02:00",
                        "name": "reviews",
                        "resourceVersion": "393057",
                      },
                      "spec": Object {
                        "gateways": undefined,
                        "hosts": Array [
                          "reviews",
                        ],
                        "http": Array [
                          Object {
                            "route": Array [
                              Object {
                                "destination": Object {
                                  "host": "reviews",
                                  "subset": "v1",
                                },
                              },
                            ],
                          },
                        ],
                        "tcp": undefined,
                      },
                    },
                  ],
                  "permissions": Object {
                    "create": false,
                    "delete": false,
                    "update": false,
                  },
                },
                "workloads": Array [],
              }
            }
            virtualServices={
              Array [
                Object {
                  "metadata": Object {
                    "creationTimestamp": "2018-07-02T13:44:01+02:00",
                    "name": "reviews",
                    "resourceVersion": "393057",
                  },
                  "spec": Object {
                    "gateways": undefined,
                    "hosts": Array [
                      "reviews",
                    ],
                    "http": Array [
                      Object {
                        "route": Array [
                          Object {
                            "destination": Object {
                              "host": "reviews",
                              "subset": "v1",
                            },
                          },
                        ],
                      },
                    ],
                    "tcp": undefined,
                  },
                },
              ]
            }
          />
        </ErrorBoundaryWithMessage>
      </ForwardRef>
      <ForwardRef
        eventKey={2}
        title={
          <React.Fragment>
            Destination Rules (
            1
            )
            <span
              className="f1uxujqj"
            >
               
              <Validation
                severity="error"
              />
            </span>
          </React.Fragment>
        }
      >
        <ErrorBoundaryWithMessage
          message="One of the Destination Rules associated to this service has an invalid format"
        >
          <ServiceInfoDestinationRules
            destinationRules={
              Array [
                Object {
                  "metadata": Object {
                    "creationTimestamp": "2018-07-02T13:44:01+02:00",
                    "name": "reviews",
                    "resourceVersion": "393061",
                  },
                  "spec": Object {
                    "host": "reviews",
                    "subsets": Array [
                      Object {
                        "labels": Object {
                          "version": "v1",
                        },
                        "name": "v1",
                      },
                      Object {
                        "labels": Object {
                          "version": "v2",
                        },
                        "name": "v2",
                      },
                      Object {
                        "labels": Object {
                          "version": "v3",
                        },
                        "name": "v3",
                      },
                    ],
                    "trafficPolicy": undefined,
                  },
                },
              ]
            }
            service={
              Object {
                "apiDocumentation": Object {
                  "hasSpec": true,
                  "type": "rest",
                },
                "destinationRules": Object {
                  "items": Array [
                    Object {
                      "metadata": Object {
                        "creationTimestamp": "2018-07-02T13:44:01+02:00",
                        "name": "reviews",
                        "resourceVersion": "393061",
                      },
                      "spec": Object {
                        "host": "reviews",
                        "subsets": Array [
                          Object {
                            "labels": Object {
                              "version": "v1",
                            },
                            "name": "v1",
                          },
                          Object {
                            "labels": Object {
                              "version": "v2",
                            },
                            "name": "v2",
                          },
                          Object {
                            "labels": Object {
                              "version": "v3",
                            },
                            "name": "v3",
                          },
                        ],
                        "trafficPolicy": undefined,
                      },
                    },
                  ],
                  "permissions": Object {
                    "create": false,
                    "delete": false,
                    "update": false,
                  },
                },
                "endpoints": Array [
                  Object {
                    "addresses": Array [
                      Object {
                        "ip": "172.17.0.20",
                        "kind": "Pod",
                        "name": "reviews-v3-5f5bcb6765-hj46f",
                      },
                      Object {
                        "ip": "172.17.0.21",
                        "kind": "Pod",
                        "name": "reviews-v2-d896b68c-jnxgm",
                      },
                      Object {
                        "ip": "172.17.0.22",
                        "kind": "Pod",
                        "name": "reviews-v1-5d6696bcf7-2sls7",
                      },
                    ],
                    "ports": Array [
                      Object {
                        "name": "http",
                        "port": 9080,
                        "protocol": "TCP",
                      },
                    ],
                  },
                ],
                "health": undefined,
                "istioSidecar": true,
                "service": Object {
                  "createdAt": "2018-06-29T16:43:18+02:00",
                  "externalName": "my.database.example.com",
                  "ip": "172.30.196.248",
                  "labels": Object {
                    "app": "reviews",
                  },
                  "name": "reviews",
                  "ports": Array [
                    Object {
                      "name": "http",
                      "port": 9080,
                      "protocol": "TCP",
                    },
                  ],
                  "resourceVersion": "2652",
                  "type": "ClusterIP",
                },
                "validations": Object {
                  "destinationrule": Object {
                    "reviews": Object {
                      "checks": Array [
                        Object {
                          "message": "This subset is not found from the host",
                          "path": "spec/subsets[0]/version",
                          "severity": "error",
                        },
                        Object {
                          "message": "This subset is not found from the host",
                          "path": "spec/subsets[1]/version",
                          "severity": "error",
                        },
                      ],
                      "name": "details",
                      "objectType": "destinationrule",
                      "valid": false,
                    },
                  },
                },
                "virtualServices": Object {
                  "items": Array [
                    Object {
                      "metadata": Object {
                        "creationTimestamp": "2018-07-02T13:44:01+02:00",
                        "name": "reviews",
                        "resourceVersion": "393057",
                      },
                      "spec": Object {
                        "gateways": undefined,
                        "hosts": Array [
                          "reviews",
                        ],
                        "http": Array [
                          Object {
                            "route": Array [
                              Object {
                                "destination": Object {
                                  "host": "reviews",
                                  "subset": "v1",
                                },
                              },
                            ],
                          },
                        ],
                        "tcp": undefined,
                      },
                    },
                  ],
                  "permissions": Object {
                    "create": false,
                    "delete": false,
                    "update": false,
                  },
                },
                "workloads": Array [],
              }
            }
            validations={
              Object {
                "reviews": Object {
                  "checks": Array [
                    Object {
                      "message": "This subset is not found from the host",
                      "path": "spec/subsets[0]/version",
                      "severity": "error",
                    },
                    Object {
                      "message": "This subset is not found from the host",
                      "path": "spec/subsets[1]/version",
                      "severity": "error",
                    },
                  ],
                  "name": "details",
                  "objectType": "destinationrule",
                  "valid": false,
                },
              }
            }
          />
        </ErrorBoundaryWithMessage>
      </ForwardRef>
    </ParameterizedTabs>
  </GridItem>
</Grid>
`;
